File: solutions/go/bob/1/bob.go
// Package bob determine what Bob will reply to someone when they say something to him or ask him a question
package bob

import (
	"regexp"
	"strings"
)

// Hey should determine what Bob will reply to a remark.
func Hey(remark string) string {
	trimedStr := strings.TrimSpace(remark)

	if trimedStr == "" {
		return "Fine. Be that way!"
	}

	hasCapitalLetters := regexp.MustCompile(`[A-Z]`).MatchString(trimedStr)
	hasDowncaseLetters := regexp.MustCompile(`[a-z]`).MatchString(trimedStr)
	hasOnlyCapitalLetters := hasCapitalLetters && !hasDowncaseLetters
	isQuestion := trimedStr[len(trimedStr)-1] == '?'

	if isQuestion {
		if hasOnlyCapitalLetters {
			return "Calm down, I know what I'm doing!"
		}
		return "Sure."
	} else if hasOnlyCapitalLetters {
		return "Whoa, chill out!"
	}

	return "Whatever."
}

File: solutions/ocaml/bob/1/bob.ml
open Base

let response_for s =
  let s = String.strip s in
  if String.is_empty s then
    "Fine. Be that way!"
  else
    let is_question = String.is_suffix s ~suffix:"?" in
    let is_yelled =
      String.equal s (String.uppercase s)
      && String.exists s ~f:Char.is_uppercase
    in
    match (is_question, is_yelled) with
    | true, true -> "Calm down, I know what I'm doing!"
    | true, false -> "Sure."
    | false, true -> "Whoa, chill out!"
    | _ -> "Whatever."
